pkg/kv/txn_coord_sender.go

```diff
@@ -725,6 +725,10 @@ func (tc *TxnCoordSender) heartbeat(ctx context.Context, txnID uuid.UUID) bool {
 	// considered abandoned. If so, exit heartbeat. If ctx.Done() is not nil, then
 	// it is a cancelable Context and we skip this check and use the ctx lifetime
 	// instead of a timeout.
+	// TODO(andrei): We'd like all transactions to use a cancelable Context and
+	// get rid of hasAbandoned and txnMeta.lastUpdateNanos. The problem is
+	// internal transactions started in various RPC handlers; we should
+	// investigate using a cancelable context from GRPC.
```

Comment:
I think this might be a little confusing for everyone not involved in our DM. How about

> We should disallow non-cancelable contexts in the heartbeat goroutine and enforce that our `kv` client cancels the context when it's done. This could be possible a) by enforcing that we only use a local client (i.e. never use the grpc transport; this is mostly true right now except for perhaps some debug commands - in this case) or, if a) isn't possible, by making sure that there's always a context with a reasonable cancellation condition (for example, through the grpc connection plus the retry closure wrapper enforcing proper cancellation on top).